### DISCUSSION TITLE
Test Driven Development (TDD) Mode

### DISCUSSION THREAD

**Discussion by FoxRedRisk** [cline/cline#535] [Feature Requests]
It would be incredibly useful to have a mode where the model first creates tests, see them go red, then implement the functionality to make them go green, then refactor the code to improve the functionality.

I have found this speeds up development and also helps prompt the LLM on how to fix bugs.

**Comment by ChristianWeyer** [comment]
Very cool idea!

**Comment by Trapper4888** [comment]
I mean, can't you just ask Cline to do it that way?

**Reply by FoxRedRisk** [reply]
You can, but not consistently. 

It tends to claim it has completed the tasks without a) running the test and b) implementing the actual functionality. I want it to explicitly test red, then green, and the option to refactor or move on.

It also tends to change the test to fit current functionality rather than change functionality to pass the tests. I want it to only refactor tests when I ask it to.

It also mocks functionality when you have explicitly asked it not to. I want this to be more deterministic.

It also doesn't listen to which testing framework you are using. If I am using unittest, I don't want it to switch to pytest halfway through.

I kind of envisage a separate agent like the architect feature in Aider, which just creates and runs tests with Clinebot, viewing the output of the tests and then implementing the functionality.

**Comment by TomLucidor** [comment]
Other than Test-Driven Development, please also include Feature Driven Development (ala Cursor hacks through using documentation/requirement MD files)
See also for inspiration this https://github.com/cline/cline/discussions/420
